cirq-google/cirq_google/engine/stream_response_demux.py

```diff
@@ -0,0 +1,81 @@
+import duet
+
+from cirq_google.cloud import quantum
+
+JobPath = str
+MessageId = str
+
+
+class _StreamResponseDemux:
+    """A event demultiplexer for QuantumRunStreamResponses, as part of the async reactor pattern.
+
+    TODO(#5996) expand
+    """
+
+    def __init__(self):
+        self._subscribers: dict[JobPath, dict[MessageId, duet.AwaitableFuture]] = {}
+
+    def subscribe(
+        self, request: quantum.QuantumRunStreamRequest
+    ) -> duet.AwaitableFuture[quantum.QuantumRunStreamResponse]:
+        """Subscribes to the QuantumRunStreamResponse matching the given request.
+
+        The request must be unique among all calls to this method, with message ID as the
+        identifier.
+
+        Returns:
+            A future for the response to be fulfilled when publish is called.
+        Raises:
+            ValueError: If there is already a subscriber for the given request.
+        """
+
+        request_type = request.WhichOneOf('request')
+        if request_type == 'create_quantum_program_and_job':
+            job_path = request.create_quantum_program_and_job.quantum_job.name
+        elif request_type == 'create_quantum_job':
+            job_path = request.create_quantum_job.quantum_job.name
+        else:  # request_type == 'get_quantum_result'
+            job_path = request.get_quantum_result.parent
+
+        if job_path not in self._subscribers:
+            self._subscribers[job_path] = {}
+
+        subscribers_by_message = self._subscribers[job_path]
+        if request.message_id in subscribers_by_message:
+            raise ValueError(f'There is another subscriber for the message ID {request.message_id}')
+
+        response_future = duet.AwaitableFuture[quantum.QuantumRunStreamResponse]()
```

Comment:
Another thing we should do is hook up cancellation. We can do this with `response_future.add_done_callback` to add a callback that will be invoked when the future completes, then checking it it was cancelled and if so send the appropriate cancellation request over the stream.